// @cityofzion/neon-api/lib/provider/neonDB/core.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { logging, u, wallet } from "@cityofzion/neon-core";
import axios from "axios";
import { settings as internalSettings } from "../../settings";
import { filterHttpsOnly, findGoodNodesFromHeight, getBestUrl } from "../common";
const log = logging.default("api");
/**
 * Returns an appropriate RPC endpoint retrieved from a neonDB endpoint.
 * @param url - URL of a neonDB service.
 * @returns URL of a good RPC endpoint.
 */
export function getRPCEndpoint(url) {
    return __awaiter(this, void 0, void 0, function* () {
        const response = yield axios.get(url + "/v2/network/nodes");
        const data = response.data.nodes;
        let nodes = data
            .filter(d => d.status)
            .map(d => ({ height: d.block_height, url: d.url }));
        if (internalSettings.httpsOnly) {
            nodes = filterHttpsOnly(nodes);
        }
        const goodNodes = findGoodNodesFromHeight(nodes);
        const bestRPC = yield getBestUrl(goodNodes);
        return bestRPC;
    });
}
/**
 * Get balances of NEO and GAS for an address
 * @param url - URL of a neonDB service.
 * @param address - Address to check.
 * @return  Balance of address
 */
export function getBalance(url, address) {
    return __awaiter(this, void 0, void 0, function* () {
        const response = yield axios.get(url + "/v2/address/balance/" + address);
        const data = response.data;
        const bal = new wallet.Balance({ net: url, address });
        if (data.NEO.balance > 0) {
            bal.addAsset("NEO", data.NEO);
        }
        if (data.GAS.balance > 0) {
            bal.addAsset("GAS", data.GAS);
        }
        log.info(`Retrieved Balance for ${address} from neonDB ${url}`);
        return bal;
    });
}
/**
 * Get amounts of available (spent) and unavailable claims.
 * @param url - URL of a neonDB service.
 * @param address - Address to check.
 * @return An object with available and unavailable GAS amounts.
 */
export function getClaims(url, address) {
    return __awaiter(this, void 0, void 0, function* () {
        const response = yield axios.get(url + "/v2/address/claims/" + address);
        const data = response.data;
        const claims = data.claims.map(c => {
            return {
                claim: new u.Fixed8(c.claim || 0).div(100000000),
                index: c.index,
                txid: c.txid,
                start: c.start || 0,
                end: c.end || 0,
                value: c.value
            };
        });
        log.info(`Retrieved Claims for ${address} from neonDB ${url}`);
        return new wallet.Claims({
            net: url,
            address,
            claims
        });
    });
}
/**
 * Gets the maximum amount of gas claimable after spending all NEO.
 * @param url - URL of a neonDB service.
 * @param address - Address to check.
 * @return An object with available and unavailable GAS amounts.
 */
export function getMaxClaimAmount(url, address) {
    return __awaiter(this, void 0, void 0, function* () {
        const response = yield axios.get(url + "/v2/address/claims/" + address);
        const data = response.data;
        log.info(`Retrieved maximum amount of gas claimable after spending all NEO for ${address} from neonDB ${url}`);
        return new u.Fixed8(data.total_claim + data.total_unspent_claim).div(100000000);
    });
}
/**
 * Get transaction history for an account
 * @param url - URL of a neonDB service.
 * @param address - Address to check.
 * @return a list of PastTransaction
 */
export function getTransactionHistory(url, address) {
    return __awaiter(this, void 0, void 0, function* () {
        const response = yield axios.get(url + "/v2/address/history/" + address);
        const data = response.data;
        log.info(`Retrieved History for ${address} from neonDB ${url}`);
        return data.history.map(rawTx => {
            return {
                change: {
                    NEO: new u.Fixed8(rawTx.NEO || 0),
                    GAS: new u.Fixed8(rawTx.GAS || 0)
                },
                blockHeight: rawTx.block_index,
                txid: rawTx.txid
            };
        });
    });
}
/**
 * Get the current height of the light wallet DB
 * @param url - URL of a neonDB service.
 * @return Current height.
 */
export function getHeight(url) {
    return __awaiter(this, void 0, void 0, function* () {
        const response = yield axios.get(url + "/v2/block/height");
        return parseInt(response.data.block_height, 10);
    });
}
