// @cityofzion/neon-core/lib/rpc/Network.js
import { compareObject, compareUnsortedPlainArrays } from "../helper";
import logger from "../logging";
import Protocol from "./Protocol";
const log = logger("protocol");
/**
 * Network interface representing a NEO blockchain network.
 * This inherits from the network.protocol file used in the C# implementation and adds in additional configurations.
 * @param config NetworkLike JS object
 */
export class Network {
    constructor(config = {}, name = null) {
        this.name = config.Name || config.name || name || "RandomNet";
        const protocolLike = Object.assign({}, config.protocol || config.ProtocolConfiguration || {});
        this.protocol = new Protocol(protocolLike);
        this.nodes = config.Nodes || config.nodes || [];
        this.extra = Object.assign({}, config.ExtraConfiguration || config.extra || {});
    }
    /**
     * Exports the class as a JSON format.
     */
    export() {
        return {
            ProtocolConfiguration: this.protocol.export(),
            Name: this.name,
            ExtraConfiguration: this.extra,
            Nodes: this.nodes
        };
    }
    equals(other) {
        return (this.name === other.name &&
            this.protocol.equals(other.protocol || {}) &&
            compareUnsortedPlainArrays(this.nodes, other.nodes || []) &&
            compareObject(this.extra, other.extra || {}));
    }
}
export default Network;
