// @cityofzion/neon-core/lib/tx/transaction/ClaimTransaction.js
import { ASSET_ID, TX_VERSION } from "../../consts";
import { Fixed8 } from "../../u";
import { getScriptHashFromAddress } from "../../wallet";
import { TransactionInput, TransactionOutput } from "../components";
import { BaseTransaction } from "./BaseTransaction";
import { serializeArrayOf } from "./main";
const MAX_CLAIMS_LENGTH = 255;
/**
 * Transaction used for claiming GAS. Do note that GAS Claims can only occur for spent coins.
 * Constructed with a list of claims and a TransactionOutput representing the GAS claimed.
 */
export class ClaimTransaction extends BaseTransaction {
    constructor(obj = {}) {
        super(Object.assign({ version: TX_VERSION.CLAIM }, obj));
        this.type = 0x02;
        this.claims = Array.isArray(obj.claims)
            ? obj.claims.slice(0, MAX_CLAIMS_LENGTH).map(c => new TransactionInput(c))
            : [];
    }
    static deserializeExclusive(ss, tx) {
        const out = {
            claims: []
        };
        const claimLength = ss.readVarInt();
        for (let i = 0; i < claimLength; i++) {
            out.claims.push(TransactionInput.fromStream(ss));
        }
        return Object.assign(tx, out);
    }
    static fromClaims(claim) {
        const totalClaim = claim.claims.reduce((p, c) => {
            return p.add(c.claim);
        }, new Fixed8(0));
        const outputs = [
            new TransactionOutput({
                assetId: ASSET_ID.GAS,
                value: totalClaim,
                scriptHash: getScriptHashFromAddress(claim.address)
            })
        ];
        const claims = claim.claims.map(c => ({
            prevHash: c.txid,
            prevIndex: c.index
        }));
        return new ClaimTransaction({ outputs, claims });
    }
    get exclusiveData() {
        return { claims: this.claims };
    }
    get fees() {
        return 0;
    }
    addClaims(claim) {
        if (this.claims.length !== 0) {
            throw new Error(`This transaction already has claims!`);
        }
        const totalClaim = claim.claims.reduce((p, c) => {
            return p.add(c.claim);
        }, new Fixed8(0));
        this.outputs.push(new TransactionOutput({
            assetId: ASSET_ID.GAS,
            value: totalClaim,
            scriptHash: getScriptHashFromAddress(claim.address)
        }));
        this.claims = claim.claims.map(c => new TransactionInput({
            prevHash: c.txid,
            prevIndex: c.index
        }));
        return this;
    }
    serializeExclusive() {
        return serializeArrayOf(this.claims);
    }
    equals(other) {
        if (this.type !== other.type) {
            return false;
        }
        if (other instanceof ClaimTransaction) {
            return this.hash === other.hash;
        }
        return this.hash === new ClaimTransaction(other).hash;
    }
    export() {
        return Object.assign(super.export(), {
            claims: this.claims.map(c => c.export())
        });
    }
}
export default ClaimTransaction;
