// @cityofzion/neon-core/lib/tx/transaction/EnrollmentTransaction.js
import { TX_VERSION } from "../../consts";
import { DEFAULT_SYSFEE } from "../../consts";
import { getPublicKeyEncoded, getPublicKeyUnencoded } from "../../wallet/core";
import { BaseTransaction } from "./BaseTransaction";
import TransactionType from "./TransactionType";
export class EnrollmentTransaction extends BaseTransaction {
    constructor(obj = {}) {
        super(Object.assign({ version: TX_VERSION.ENROLLMENT }, obj));
        this.type = TransactionType.EnrollmentTransaction;
        this.publicKey = obj.publicKey || "";
    }
    static deserializeExclusive(ss, tx) {
        const hexStrPrefix = ss.read(1);
        const prefix = parseInt(hexStrPrefix, 16);
        let pKey = "";
        // Compressed public keys.
        if (prefix === 0x02 || prefix === 0x03) {
            pKey = ss.read(32);
        }
        else if (prefix === 0x04) {
            pKey = ss.read(64);
        }
        else if (prefix === 0x00) {
            // infinity case.
            return Object.assign(tx, { publicKey: "" });
        }
        else {
            throw new Error("Prefix not recognised for public key");
        }
        pKey = hexStrPrefix + pKey;
        const publicKey = getPublicKeyUnencoded(pKey);
        return Object.assign(tx, { publicKey });
    }
    get exclusiveData() {
        return { publicKey: this.publicKey };
    }
    get fees() {
        return DEFAULT_SYSFEE.enrollmentTransaction;
    }
    serializeExclusive() {
        if (this.publicKey === "") {
            return "00";
        }
        return getPublicKeyEncoded(this.publicKey);
    }
    export() {
        return Object.assign(super.export(), {
            publicKey: this.publicKey
        });
    }
    equals(other) {
        if (this.type !== other.type) {
            return false;
        }
        if (other instanceof EnrollmentTransaction) {
            return this.hash === other.hash;
        }
        return this.hash === new EnrollmentTransaction(other).hash;
    }
}
export default EnrollmentTransaction;
