// @cityofzion/neon-core/lib/tx/transaction/InvocationTransaction.js
import { TX_VERSION } from "../../consts";
import { Fixed8, fixed82num, num2VarInt } from "../../u";
import { BaseTransaction } from "./BaseTransaction";
/**
 * Transaction used for invoking smart contracts through a VM script.
 * Can also be used to transfer UTXO assets.
 */
export class InvocationTransaction extends BaseTransaction {
    constructor(obj = {}) {
        super(Object.assign({ version: TX_VERSION.INVOCATION }, obj));
        this.type = 0xd1;
        this.script = obj.script || "";
        this.gas = new Fixed8(obj.gas);
    }
    static deserializeExclusive(ss, tx) {
        const script = ss.readVarBytes();
        const version = parseInt(ss.str.substr(2, 2), 16);
        const gas = version >= 1 ? fixed82num(ss.read(8)) : 0;
        return Object.assign(tx, { script, gas });
    }
    get exclusiveData() {
        return { gas: this.gas, script: this.script };
    }
    get fees() {
        return this.gas.toNumber();
    }
    serializeExclusive() {
        let out = num2VarInt(this.script.length / 2);
        out += this.script;
        if (this.version >= 1) {
            out += this.gas.toReverseHex();
        }
        return out;
    }
    equals(other) {
        if (this.type !== other.type) {
            return false;
        }
        if (other instanceof InvocationTransaction) {
            return this.hash === other.hash;
        }
        return this.hash === new InvocationTransaction(other).hash;
    }
    export() {
        return Object.assign(super.export(), {
            script: this.script,
            gas: this.gas.toNumber()
        });
    }
}
export default InvocationTransaction;
