// @cityofzion/neon-core/lib/tx/components/TransactionInput.js
import { num2hexstring, reverseHex, StringStream } from "../../u";
/**
 * A reference to a TransactionOutput in another confirmed transaction.
 * This is used to denote UTXO that will be spent in this transaction.
 */
export class TransactionInput {
    static deserialize(hex) {
        const ss = new StringStream(hex);
        return this.fromStream(ss);
    }
    static fromStream(ss) {
        const prevHash = reverseHex(ss.read(32));
        const prevIndex = parseInt(reverseHex(ss.read(2)), 16);
        return new TransactionInput({ prevHash, prevIndex });
    }
    constructor(obj) {
        if (!obj || obj.prevHash === undefined || obj.prevIndex === undefined) {
            throw new Error("TransactionInput requires prevHash and prevIndex fields");
        }
        this.prevHash = obj.prevHash;
        this.prevIndex = obj.prevIndex;
    }
    serialize() {
        return (reverseHex(this.prevHash) + reverseHex(num2hexstring(this.prevIndex, 2)));
    }
    export() {
        return {
            prevHash: this.prevHash,
            prevIndex: this.prevIndex
        };
    }
    equals(other) {
        return (this.prevHash === other.prevHash && this.prevIndex === other.prevIndex);
    }
}
export default TransactionInput;
